src/ja/server/scheduler/default_policies.py

```diff
@@ -29,51 +45,129 @@ class DefaultJobDistributionPolicyBase(JobDistributionPolicy, ABC):
+            if self._cost_func.calculate_cost(job) > self._cost_func.preempting_threshold:
+                return None
+
+        after_allocation = machine.resources.free_resources - self._get_job_allocation(job.job)
+        return (after_allocation.cpu_threads * self._cpu_threads_w + after_allocation.memory * self._memory_w, [])
 
 
 class DefaultBlockingDistributionPolicy(DefaultJobDistributionPolicyBase):
     """
     Default distribution policy for blocking jobs.
     """
     def _assign_machine_cost(self,
-                             job: Job,
+                             job: DatabaseJobEntry,
                              machine: WorkMachine,
-                             existing_jobs: List[Job]) -> Optional[Tuple[int, List[Job]]]:
-        pass
+                             existing_jobs: List[DatabaseJobEntry]) -> Optional[Tuple[float, List[Job]]]:
+        if not self._check_machine_feasible(job.job, machine, free_only=False):
+            return None
+        return (len(existing_jobs), [])
 
 
 class DefaultPreemptiveDistributionPolicy(DefaultJobDistributionPolicyBase):
+    _inverse_multiplier = 10.0
     """
     Default distribution policy for preemptive jobs.
     """
+    def __init__(self, cost_function: CostFunction):
+        """!
+        Initialize the default preemptive distribution policy.
+
+        @param cost_function The cost function to assing priorities to tasks.
+        """
+        self._cost_func = cost_function
+
     def _assign_machine_cost(self,
-                             job: Job,
+                             job: DatabaseJobEntry,
                              machine: WorkMachine,
-                             existing_jobs: List[Job]) -> Optional[Tuple[int, List[Job]]]:
-        pass
+                             existing_jobs: List[DatabaseJobEntry]) -> Optional[Tuple[float, List[Job]]]:
+        if not self._check_machine_feasible(job.job, machine, free_only=False):
+            return None
+
+        sorted_jobs = sorted(existing_jobs, key=lambda je: (self._cost_func.calculate_cost(je)), reverse=True)
+        sorted_jobs = list(filter(lambda je: (not je.job.scheduling_constraints.is_preemptible), sorted_jobs))
+        job_allocation = self._get_job_allocation(job.job)
+        allocation = deepcopy(machine.resources.free_resources)
+
+        total_cost: float = 0
+        to_preempt: List[Job] = []
+        while (allocation - job_allocation).is_negative():
+            if not sorted_jobs:
+                return None
+
+            next_job = sorted_jobs[0]
+            sorted_jobs = sorted_jobs[1:]
+            next_job_cost = self._cost_func.calculate_cost(next_job)
+            if next_job_cost <= self._cost_func.preempting_threshold:
+                # We cannot pause enough jobs on this machine
+                return None
+
+            total_cost += self._inverse_multiplier / (next_job_cost - self._cost_func.preempting_threshold)
```

Comment:
I do not think that just taking the maximum (actually minimum) cost of the preempted jobs is the best approach. Consider a case where we have a lot of medium priority jobs, and we have multiple machines where the minimum priorities are almost the same.

In that case, I think it also matters how many jobs we want to pause. Hence, I made it a sum.
However, maybe there is a better formula so that the lowest-priority job gets more "weight" compared to the higher-priority jobs, while still allowing us to take number of jobs into account.

Reply:
>In that case, I think it also matters how many jobs we want to pause.

Why?
Preemption is essentially the same problem as regular scheduling: to pick which jobs to run with your resources.
When we schedule jobs we go strictly by priority and not priority/resources.
We discussed this problem before: should we schedule 10 low-priority jobs or 1 high-priority job?
And what we agreed on was that we should just always try to schedule the highest-priority jobs first and then try to run lower-priority jobs with the leftover resources.
Otherwise you would not follow the user-provided priorities and get a bias against large jobs.
Applied to preemption the strategy would be to pause the lowest-priority jobs until you have enough resources and to then consider the priority of only the highest-priority job that you are pausing.

Reply:
I agree with the overall idea, I am just saying we also need to take the number of paused jobs if the highest-priority job-to-be-paused would be almost the same for two machines.

Like, if you have a machine with 1 Medium job, and another machine with 1 medium and 10 low-priority jobs, the job(s) on which machine would you prefer to preempt?

Reply:
I think it doesn't matter which jobs we pause if the priorities are almost the same anyways.
The number of jobs doesn't matter.
What matters is the amount of resources that goes towards jobs with high priority.

Also, in the scenario you're describing: it would depend on which jobs would need to be preempted to run the preempting job.

Reply:
In both scenarios, you need to pause all the jobs. (the machine with many jobs, the jobs there are all small)

Reply:
I changed the formula a little bit and I think that right now it does what both of us want.